src/components/fields/valueType/ValueTypeRenderer.tsx

```diff
@@ -0,0 +1,125 @@
+import i18n from '@dhis2/d2-i18n'
+import { CheckboxFieldFF, InputFieldFF, TextAreaFieldFF } from '@dhis2/ui'
+import React, { useMemo } from 'react'
+import { FieldRenderProps } from 'react-final-form'
+import { ModelSingleSelectField } from '../../metadataFormControls/ModelSingleSelect'
+import { ValueType } from './types'
+
+type CommonFieldProps = {
+    label?: string
+    required?: boolean
+}
+
+type FinalFormFieldProps = Pick<FieldRenderProps<string>, 'input' | 'meta'>
+
+/**
+ * A component that renders an underlying Field based on the valueType
+ *
+ * Note that this is not bound to a form - but it follows the API of ReactFinalForm
+ *  eg. using input and meta
+ * The output of all fields is a string (eg. onChange) should be a string
+ */
+export function ValueTypeRenderer(
+    props: CommonFieldProps & FinalFormFieldProps & { valueType: ValueType }
+) {
+    const valueType = props.valueType
+    const value = props.input.value
+
+    // ModelSingleSelects needs value to be an object, but the value from field is an id
+    // so keep this stable
+    const stableIdValue = useMemo(() => {
+        return { id: value }
+    }, [value])
+
+    const stableInput = useMemo(() => {
+        return { ...props.input, value: stableIdValue }
+    }, [props.input, stableIdValue])
+
+    if (valueType === 'BOOLEAN' || valueType === 'TRUE_ONLY') {
+        return (
+            <CheckboxFieldFF
+                {...props}
+                type="checkbox"
+                input={{
+                    ...props.input,
+                    checked: value === 'true',
+                    onChange: (e) => {
+                        const falseValue =
+                            valueType === 'BOOLEAN' ? 'false' : undefined
+                        props.input.onChange(
+                            e.target.checked ? 'true' : falseValue
+                        )
+                    },
+                }}
+            />
+        )
+    }
+
+    if (valueType === 'TEXT') {
+        return <InputFieldFF {...props} />
+    }
+    if (valueType === 'LONG_TEXT') {
+        return <TextAreaFieldFF {...props} />
+    }
+
+    if (valueType === 'NUMBER' || valueType === 'INTEGER') {
+        return (
+            <InputFieldFF
+                {...props}
+                input={{
+                    ...props.input,
+                    type: 'number',
+                }}
+            />
+        )
+    }
+
+    if (valueType === 'GEOJSON') {
+        return (
+            <TextAreaFieldFF
+                {...props}
+                resize="both"
+                helpText={i18n.t('Please enter a GeoJSON value')}
+            />
+        )
+    }
+
+    if (valueType === 'FILE_RESOURCE') {
+        return (
+            <ModelSingleSelectField
```

Comment:
Note that this does not allow you to upload a file in the field. I'm not sure if we even want that. The old app just renders regular input field... So should be an improvement by rendering a list of available fileresources.